p8e-util/src/main/kotlin/io/p8e/crypto/SmartKeySigner.kt

```diff
@@ -45,64 +48,119 @@ import java.security.spec.X509EncodedKeySpec
  * operating in a public, private, or hybrid cloud.
  */
 
-class SmartKeySigner(
-    private val appApiKey: String,
-    private val keyUuid: String
-): SignerImpl {
+class SmartKeySigner: SignerImpl {
 
     init {
-        val client = ApiClient()
-        client.setBasicAuthString(appApiKey)
-        Configuration.setDefaultApiClient(client)
-
-        val authResponse = AuthenticationApi().authorize()
-        val auth = client.getAuthentication("bearerToken") as ApiKeyAuth
-        auth.apiKey = authResponse.accessToken
-        auth.apiKeyPrefix = "Bearer"
+        Security.addProvider(BouncyCastleProvider())
     }
 
-    override fun sign(data: String): Signature = sign(data.toByteArray())
+    companion object {
+        // Algo must match Provenance-object-store
+        val SIGN_ALGO = "SHA512withECDSA"
+        val PROVIDER = BouncyCastleProvider.PROVIDER_NAME
+    }
+
+    var signature: Signature? = null
+
+    private var signatureRequest: SignRequest? = null
+    private var verifying: Boolean = false
+    private var keyUuid: String? = null
+
+    private val cachedData = mutableListOf<ByteArray>()
+
+    fun instance(keyUuid: String): SmartKeySigner {
+        this.keyUuid = keyUuid
+        return this
+    }
+
+    /**
+     * Using the local java security signature instance to verify data.
+     */
+    override fun initVerify(publicKey: PublicKey) {
+        signature = Signature.getInstance(SIGN_ALGO, PROVIDER).apply { initVerify(publicKey) }
+        verifying = true
+    }
+
+    /**
+     * Using SmartKey to sign data.
+     */
+    override fun initSign() {
+        signatureRequest = SignRequest().hashAlg(DigestAlgorithm.SHA512).deterministicSignature(true)
+        verifying = false
+    }
+
+    override fun update(data: ByteArray, off: Int, len: Int) {
+        if(!verifying) {
+            signatureRequest?.data(data.copyOfRange(off, off+len))
+        } else {
+            /**
+             * Large data dump from the InputStream drops in here, we only care about
+             * the last piece of data to verify against. If we continue to update the
+             * signature object with every single data that pass through the verification
+             * returns a false negative.
+             *
+             * Signature.getInstance is a 1:1 - meaning new instance for each piece of data.
+             */
+
+            cachedData.add(data.copyOfRange(off, off+len))
```

Comment:
Not a big fan of this, looking into ways to improve this, ideas appreciated. Essentially, when we push to OS, and then fetch the data for verification, the data that we signed for gets appended to the end of the input stream, we don't care for anything before and at the same time we don't want to mess up our local signature object during data update (Hence, L#111).